File: sdks/typescript/src/v1/client/features/workflows.ts
import { BaseWorkflowDeclaration, WorkflowDefinition } from '@hatchet/v1';
import type { LegacyWorkflow } from '@hatchet/legacy/legacy-transformer';
import { isLegacyWorkflow, warnLegacyWorkflow } from '@hatchet/legacy/legacy-transformer';
import { isValidUUID } from '@util/uuid';
import {
  WorkflowPauseScheduledCronRunQueueBehavior,
  WorkflowUpdateRequest,
} from '@hatchet/clients/rest/generated/data-contracts';
import { HatchetClient } from '../client';
import { Duration, durationToString } from '../duration';

export { WorkflowPauseScheduledCronRunQueueBehavior };

export type PauseWorkflowOpts = {
  /** How long runs stay queued while the workflow is paused before they are dropped. */
  queueTTL: Duration;
  /** The behavior of cron runs triggered while the workflow is paused. Defaults to QUEUE. */
  pausedWorkflowCronRunQueueBehavior?: WorkflowPauseScheduledCronRunQueueBehavior;
  /** The behavior of scheduled runs triggered while the workflow is paused. Defaults to QUEUE. */
  pausedWorkflowScheduledRunQueueBehavior?: WorkflowPauseScheduledCronRunQueueBehavior;
};

export const workflowNameString = (
  workflow: string | WorkflowDefinition | BaseWorkflowDeclaration<any, any> | LegacyWorkflow
) => {
  if (typeof workflow === 'string') {
    return workflow;
  }
  if (typeof workflow === 'object' && 'name' in workflow) {
    return workflow.name as string;
  }
  if (isLegacyWorkflow(workflow)) {
    warnLegacyWorkflow();
    return workflow.id;
  }
  throw new Error(
    'Invalid workflow: must be a string, Workflow object, or WorkflowDefinition object'
  );
};

/**
 * The workflows client is a client for managing workflows programmatically within Hatchet.
 *
 * NOTE: that workflows are the declaration, not the individual runs. If you're looking for runs, use the RunsClient instead.
 *
 */
export class WorkflowsClient {
  api: HatchetClient['api'];
  tenantId: string;
  // Add a cache for workflows
  private workflowCache: Map<string, { workflow: any; expiry: number }> = new Map();
  // Default cache TTL: 5 minutes (in ms)
  private cacheTTL: number = 5 * 60 * 1000;

  constructor(client: HatchetClient, cacheTTL?: number) {
    this.api = client.api;
    this.tenantId = client.tenantId;
    // Allow custom cache TTL if provided
    if (cacheTTL !== undefined) {
      this.cacheTTL = cacheTTL;
    }
  }

  /**
   * Gets the workflow ID from a workflow name, ID, or object.
   * If the input is not a valid UUID, it will look up the workflow by name.
   * @param workflow - The workflow name, ID, or object.
   * @returns The workflow ID as a string.
   */
  async getWorkflowIdFromName(
    workflow: string | WorkflowDefinition | BaseWorkflowDeclaration<any, any> | LegacyWorkflow
  ): Promise<string> {
    const str = (() => {
      if (typeof workflow === 'string') {
        return workflow;
      }

      if (typeof workflow === 'object' && 'name' in workflow) {
        return workflow.name;
      }

      throw new Error(
        'Invalid workflow: must be a string, Workflow object, or WorkflowDefinition object'
      );
    })();

    if (!isValidUUID(str)) {
      const wf = await this.get(str);
      if (!wf) {
        throw new Error('Invalid workflow ID: must be a valid UUID');
      }
      return wf.metadata.id;
    }

    return str;
  }

  /**
   * Get a workflow by its name, ID, or object.
   * @param workflow - The workflow name, ID, or object.
   * @returns A promise that resolves to the workflow.
   */
  async get(workflow: string | BaseWorkflowDeclaration<any, any> | LegacyWorkflow) {
    // Get workflow name string
    const name = workflowNameString(workflow);

    // Check cache first
    const cached = this.workflowCache.get(name);
    if (cached && cached.expiry > Date.now()) {
      return cached.workflow as NonNullable<
        Awaited<ReturnType<typeof this.api.workflowList>>['data']['rows']
      >[number];
    }

    // If not in cache or expired, fetch from API
    try {
      // Since the API only supports listing with a name filter,
      // we'll use the list endpoint with a name filter
      const { data } = await this.api.workflowList(this.tenantId, {
        name,
      });

      if (data && data.rows && data.rows.length > 0) {
        const wf = data.rows.find((row) => row.name === name) ?? data.rows[0];

        this.workflowCache.set(name, {
          workflow: wf,
          expiry: Date.now() + this.cacheTTL,
        });

        return wf;
      }

      throw new Error(`Workflow with name ${name} not found`);
    } catch (error) {
      // Clear cache on error
      this.workflowCache.delete(name);
      throw error;
    }
  }

  /**
   * List all workflows in the tenant.
   * @param opts - The options for the list operation.
   * @returns A promise that resolves to the list of workflows.
   */
  async list(opts?: Parameters<typeof this.api.workflowList>[1]) {
    const { data } = await this.api.workflowList(this.tenantId, opts);
    return data;
  }

  /**
   * Delete a workflow by its name, ID, or object.
   * @param workflow - The workflow name, ID, or object.
   * @returns A promise that resolves to the deleted workflow.
   */
  async delete(workflow: string | BaseWorkflowDeclaration<any, any> | LegacyWorkflow) {
    const name = workflowNameString(workflow);

    try {
      // Get the workflow first to find its ID
      const workflowObj = await this.get(name);

      if (!workflowObj || !workflowObj.metadata || !workflowObj.metadata.id) {
        throw new Error(`Could not find workflow with name ${name}`);
      }

      const { data } = await this.api.workflowDelete(workflowObj.metadata.id);

      // Remove from cache after deletion
      this.workflowCache.delete(name);

      return data;
    } catch (error) {
      // Clear cache on error
      this.workflowCache.delete(name);
      throw error;
    }
  }

  /**
   * Pause a workflow. While paused, new runs of the workflow are queued but not started.
   * @param workflow - The workflow name, ID, or object.
   * @param opts - The options for the pause operation.
   * @returns A promise that resolves to the updated workflow.
   */
  async pause(
    workflow: string | BaseWorkflowDeclaration<any, any> | LegacyWorkflow,
    opts: PauseWorkflowOpts
  ) {
    return this.update(workflow, {
      pause: {
        action: 'pause',
        pausedWorkflowQueueTTL: durationToString(opts.queueTTL),
        pausedWorkflowCronRunQueueBehavior:
          opts.pausedWorkflowCronRunQueueBehavior ??
          WorkflowPauseScheduledCronRunQueueBehavior.QUEUE,
        pausedWorkflowScheduledRunQueueBehavior:
          opts.pausedWorkflowScheduledRunQueueBehavior ??
          WorkflowPauseScheduledCronRunQueueBehavior.QUEUE,
      },
    });
  }

  /**
   * Unpause a workflow.
   * @param workflow - The workflow name, ID, or object.
   * @returns A promise that resolves to the updated workflow.
   */
  async unpause(workflow: string | BaseWorkflowDeclaration<any, any> | LegacyWorkflow) {
    return this.update(workflow, { pause: { action: 'unpause' } });
  }

  private async update(
    workflow: string | BaseWorkflowDeclaration<any, any> | LegacyWorkflow,
    request: WorkflowUpdateRequest
  ) {
    const name = workflowNameString(workflow);
    const workflowId = await this.getWorkflowIdFromName(name);

    try {
      const { data } = await this.api.workflowUpdate(workflowId, request);
      this.workflowCache.delete(name);
      this.workflowCache.delete(data.name);
      return data;
    } catch (error) {
      this.workflowCache.delete(name);
      throw error;
    }
  }
}

File: sdks/go/features/workflows.go
package features

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/cockroachdb/errors"
	"github.com/google/uuid"

	"github.com/hatchet-dev/hatchet/pkg/client/rest"
	"github.com/hatchet-dev/hatchet/pkg/repository/cache"
)

// WorkflowsClient provides methods for interacting with workflows
type WorkflowsClient struct {
	api      *rest.ClientWithResponses
	tenantId uuid.UUID
	cache    *cache.Cache
}

// NewWorkflowsClient creates a new WorkflowsClient
func NewWorkflowsClient(
	api *rest.ClientWithResponses,
	tenantId string,
) *WorkflowsClient {
	tenantIdUUID := uuid.MustParse(tenantId)

	// Create a cache with the specified TTL
	workflowCache := cache.New(time.Minute * 5)

	return &WorkflowsClient{
		api:      api,
		tenantId: tenantIdUUID,
		cache:    workflowCache,
	}
}

// Get retrieves a workflow by its ID or name.
func (w *WorkflowsClient) Get(ctx context.Context, workflowName string) (*rest.Workflow, error) {
	// Try to get the workflow from cache first
	cacheKey := workflowName
	cachedWorkflow, found := w.cache.Get(cacheKey)
	if found {
		return cachedWorkflow.(*rest.Workflow), nil
	}

	// FIXME: this is a hack to get the workflow by name
	resp, err := w.api.WorkflowListWithResponse(
		ctx,
		w.tenantId,
		&rest.WorkflowListParams{
			Name: &workflowName,
		},
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get workflow")
	}

	if resp.JSON200 == nil || len(*resp.JSON200.Rows) == 0 {
		return nil, fmt.Errorf("workflow with name %s not found", workflowName)
	}

	workflow := (*resp.JSON200.Rows)[0]

	// Update cache
	w.cache.Set(cacheKey, &workflow)

	return &workflow, nil
}

// GetId retrieves a workflow by its name.
func (w *WorkflowsClient) GetId(ctx context.Context, workflowName string) (uuid.UUID, error) {
	workflow, err := w.Get(ctx, workflowName)
	if err != nil {
		return uuid.Nil, errors.Wrap(err, "failed to get workflow ID")
	}

	return uuid.MustParse(workflow.Metadata.Id), nil
}

// List retrieves all workflows for the tenant with optional filtering parameters.
func (w *WorkflowsClient) List(ctx context.Context, opts *rest.WorkflowListParams) (*rest.WorkflowList, error) {
	resp, err := w.api.WorkflowListWithResponse(
		ctx,
		w.tenantId,
		opts,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to list workflows")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Delete removes a workflow by its ID or name.
func (w *WorkflowsClient) Delete(ctx context.Context, workflowName string) (*rest.WorkflowDeleteResponse, error) {
	// FIXME: this is a hack to get the workflow by name
	workflowId, err := w.GetId(ctx, workflowName)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get workflow ID")
	}

	resp, err := w.api.WorkflowDeleteWithResponse(
		ctx,
		workflowId,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to delete workflow")
	}

	if err := validateStatusCodeResponse(resp.StatusCode(), resp.Body); err != nil {
		return nil, err
	}

	// Remove from cache after deletion
	w.cache.Set(workflowName, nil)

	return resp, nil
}

// PauseWorkflowOpts contains the configuration for pausing a workflow.
type PauseWorkflowOpts struct {
	// QueueTTL is how long runs stay queued while the workflow is paused before they are dropped.
	QueueTTL time.Duration

	// (optional) CronRunQueueBehavior is the behavior of cron runs triggered while the workflow is paused. Defaults to QUEUE.
	CronRunQueueBehavior rest.WorkflowPauseScheduledCronRunQueueBehavior

	// (optional) ScheduledRunQueueBehavior is the behavior of scheduled runs triggered while the workflow is paused. Defaults to QUEUE.
	ScheduledRunQueueBehavior rest.WorkflowPauseScheduledCronRunQueueBehavior
}

func queueBehaviorOrDefault(behavior rest.WorkflowPauseScheduledCronRunQueueBehavior) rest.WorkflowPauseScheduledCronRunQueueBehavior {
	if behavior == "" {
		return rest.QUEUE
	}

	return behavior
}

// Pause pauses a workflow by its name. While paused, new runs of the workflow are queued but not started.
func (w *WorkflowsClient) Pause(ctx context.Context, workflowName string, opts PauseWorkflowOpts) (*rest.Workflow, error) {
	var pauseRequest rest.PauseWorkflowRequest

	err := pauseRequest.FromPauseWorkflowRequestPause(rest.PauseWorkflowRequestPause{
		Action:                                  rest.Pause,
		PausedWorkflowQueueTTL:                  strconv.FormatFloat(opts.QueueTTL.Seconds(), 'f', -1, 64) + "s",
		PausedWorkflowCronRunQueueBehavior:      queueBehaviorOrDefault(opts.CronRunQueueBehavior),
		PausedWorkflowScheduledRunQueueBehavior: queueBehaviorOrDefault(opts.ScheduledRunQueueBehavior),
	})
	if err != nil {
		return nil, errors.Wrap(err, "failed to build pause request")
	}

	return w.update(ctx, workflowName, rest.WorkflowUpdateRequest{Pause: &pauseRequest})
}

// Unpause unpauses a workflow by its name.
func (w *WorkflowsClient) Unpause(ctx context.Context, workflowName string) (*rest.Workflow, error) {
	var unpauseRequest rest.PauseWorkflowRequest

	err := unpauseRequest.FromPauseWorkflowRequestUnpause(rest.PauseWorkflowRequestUnpause{
		Action: rest.Unpause,
	})
	if err != nil {
		return nil, errors.Wrap(err, "failed to build unpause request")
	}

	return w.update(ctx, workflowName, rest.WorkflowUpdateRequest{Pause: &unpauseRequest})
}

func (w *WorkflowsClient) update(ctx context.Context, workflowName string, request rest.WorkflowUpdateRequest) (*rest.Workflow, error) {
	workflowId, err := w.GetId(ctx, workflowName)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get workflow ID")
	}

	resp, err := w.api.WorkflowUpdateWithResponse(ctx, workflowId, request)
	if err != nil {
		return nil, errors.Wrap(err, "failed to update workflow")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	w.cache.Set(workflowName, resp.JSON200)

	return resp.JSON200, nil
}
